pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/TransferableBlockUtilsTest.java

```diff
@@ -0,0 +1,231 @@
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.query.runtime;
+
+import com.google.common.collect.ImmutableList;
+import java.math.BigDecimal;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Random;
+import org.apache.commons.lang.RandomStringUtils;
+import org.apache.pinot.common.datablock.BaseDataBlock;
+import org.apache.pinot.common.datablock.ColumnarDataBlock;
+import org.apache.pinot.common.datablock.DataBlockUtils;
+import org.apache.pinot.common.datablock.MetadataBlock;
+import org.apache.pinot.common.datablock.RowDataBlock;
+import org.apache.pinot.common.utils.DataSchema;
+import org.apache.pinot.core.common.datablock.DataBlockBuilder;
+import org.apache.pinot.query.runtime.blocks.TransferableBlockUtils;
+import org.apache.pinot.spi.utils.ByteArray;
+import org.testng.Assert;
+import org.testng.annotations.Test;
+
+import static org.testng.Assert.assertThrows;
+
+
+public class TransferableBlockUtilsTest {
+  // TODO: Consolidate the test utils with DatablockTestUtils
+  private static final Random RANDOM = new Random();
+  private static final int ARRAY_SIZE = 5;
+  private static final int TEST_ROW_COUNT = 5;
+  private static final List<DataSchema.ColumnDataType> EXCLUDE_DATA_TYPES =
+      ImmutableList.of(DataSchema.ColumnDataType.OBJECT, DataSchema.ColumnDataType.BYTES_ARRAY);
+
+  private static DataSchema getDataSchema() {
+    DataSchema.ColumnDataType[] allDataTypes = DataSchema.ColumnDataType.values();
+    List<DataSchema.ColumnDataType> columnDataTypes = new ArrayList<DataSchema.ColumnDataType>();
+    List<String> columnNames = new ArrayList<String>();
+    for (int i = 0; i < allDataTypes.length; i++) {
+      if (!EXCLUDE_DATA_TYPES.contains(allDataTypes[i])) {
+        columnNames.add(allDataTypes[i].name());
+        columnDataTypes.add(allDataTypes[i]);
+      }
+    }
+    return new DataSchema(columnNames.toArray(new String[0]),
+        columnDataTypes.toArray(new DataSchema.ColumnDataType[0]));
+  }
+
+  public static boolean randomlySettingNull(int percentile) {
+    return RANDOM.nextInt(100) >= (100 - percentile);
+  }
+
+  public static List<Object[]> getRandomRows(DataSchema dataSchema, int numRows, int nullPercentile) {
```

Comment:
it's nice not to have randomness in tests, if an error happens due to a specific randomness generator we won't be able to debug it. Can we make this deterministic?

Reply:
we can set a random seed. but for this particular test we are fine introduce some randomness

Reply:
this reminds me... it is a good idea we should add a print out for the row itself